jung-algorithms/src/main/java/edu/uci/ics/jung/layout/algorithms/SpringBHVisitorLayoutAlgorithm.java

```diff
@@ -0,0 +1,296 @@
+                  return;
+                }
+                double vx = p.x - p2.x;
+                double vy = p.y - p2.y;
+                double distanceSq = p.distanceSquared(p2);
+                if (distanceSq == 0) {
+                  f = f.add(random.nextDouble(), random.nextDouble());
+                } else if (distanceSq < repulsion_range_sq) {
+                  double factor = 1;
+                  f = f.add(factor * vx / distanceSq, factor * vy / distanceSq);
+                }
+              }
+            };
+        tree.visit(nodeForceObject, Optional.empty());
+        Point f = nodeForceObject.f;
+        double dlen = f.x * f.x + f.y * f.y;
+        if (dlen > 0) {
+          dlen = Math.sqrt(dlen) / 2;
+          svd.repulsiondx += f.x / dlen;
+          svd.repulsiondy += f.y / dlen;
+        }
+      }
+    } catch (ConcurrentModificationException cme) {
+      calculateRepulsion();
+    }
+  }
+
+  protected void moveNodes() {
+    Graph<N> graph = layoutModel.getGraph();
+
+    synchronized (layoutModel) {
+      try {
+        for (N node : graph.nodes()) {
+          if (layoutModel.isLocked(node)) {
+            continue;
+          }
+          SpringNodeData vd = springNodeData.getUnchecked(node);
+          if (vd == null) {
+            continue;
+          }
+          Point xyd = layoutModel.apply(node);
+          double posX = xyd.x;
+          double posY = xyd.y;
+
+          vd.dx += vd.repulsiondx + vd.edgedx;
+          vd.dy += vd.repulsiondy + vd.edgedy;
+
+          // keeps nodes from moving any faster than 5 per time unit
+          posX = posX + Math.max(-5, Math.min(5, vd.dx));
+          posY = posY + Math.max(-5, Math.min(5, vd.dy));
+
+          int width = layoutModel.getWidth();
+          int height = layoutModel.getHeight();
+
+          if (posX < 0) {
+            posX = 0;
+          } else if (posX > width) {
+            posX = width;
+          }
+          if (posY < 0) {
+            posY = 0;
+          } else if (posY > height) {
+            posY = height;
+          }
+          // after the bounds have been honored above, really set the location
+          // in the layout model
+          layoutModel.set(node, posX, posY);
+        }
+      } catch (ConcurrentModificationException cme) {
+        moveNodes();
+      }
+    }
+  }
+
+  protected static class SpringNodeData {
+    protected double edgedx;
+    protected double edgedy;
+    protected double repulsiondx;
+    protected double repulsiondy;
+
+    /** movement speed, x */
+    protected double dx;
+
+    /** movement speed, y */
+    protected double dy;
+
+    @Override
+    public String toString() {
+      return "{"
+          + "edge="
+          + Point.of(edgedx, edgedy)
+          + ", rep="
+          + Point.of(repulsiondx, repulsiondy)
+          + ", dx="
+          + dx
+          + ", dy="
+          + dy
+          + '}';
+    }
+  }
+
+  /** @return true */
+  public boolean isIncremental() {
```

Comment:
I think we can get rid of this method everywhere; it's equivalent to (or should be) `instanceof IterativeLayoutAlgorithm`.

Reply:
Sounds good. A general cleanup of all the algorithms would be nice, to give them a consistent style, at least.